Import pandas library

In [1]:
import pandas as pd
import numpy as ny

Read data file

In [2]:
df=pd.read_csv("https://raw.githubusercontent.com/austin1018/IS362_Assignment12/main/agaricus-lepiota.data", header=None
              ,names=["class","cap-shape", "cap-surface", "cap-color", "bruises","odor","gill-attachment","gill-spacing","gill-size"
                     ,"gill-color","stalk-shape","stalk-root","stalk-surface-above-ring","stalk-surface-below-ring"
                     ,"stalk-color-above-ring","stalk-color-below-ring","veil-type","veil-color","ring-number","ring-type"
                     ,"spore-print-color","population","habitat"])

In [3]:
df.head()

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
0     p         x           s         n       t    p               f   
1     e         x           s         y       t    a               f   
2     e         b           s         w       t    l               f   
3     p         x           y         w       t    p               f   
4     e         x           s         g       f    n               f   

  gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
0            c         n          k  ...                        s   
1            c         b          k  ...                        s   
2            c         b          n  ...                        s   
3            c         n          n  ...                        s   
4            w         b          k  ...                        s   

  stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                      w                      w         p          w   
1                      w                      w         p          w   
2                      w                      w         p          w   
3                      w                      w         p          w   
4                      w                      w         p          w   

  ring-number ring-type spore-print-color population habitat  
0           o         p                 k          s       u  
1           o         p                 n          n       g  
2           o         p                 n          n       m  
3           o         p                 k          s       u  
4           o         e                 n          a       g  

[5 rows x 23 columns]

Take the subset from original dataset

In [4]:
df_subset= df[["class", "odor","cap-shape"]]

Replace the value with number to prepare for showing data in scatter plot

In [5]:
df_subset["class"]=df_subset["class"].str.replace("p","1").replace("e","2")
df_subset["odor"]=df_subset["odor"].str.replace("a","1").replace("l","2").replace("c","3").replace("y","4").replace("f","5").replace("m","6").replace("n","7").replace("p","8").replace("s","9")
df_subset["cap-shape"]=df_subset["cap-shape"].str.replace("b","1").replace("c","2").replace("x","3").replace("f","4").replace("k","5").replace("s","6")

<ipython-input-5-67a172f8eb6b>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_subset["class"]=df_subset["class"].str.replace("p","1").replace("e","2")
<ipython-input-5-67a172f8eb6b>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_subset["odor"]=df_subset["odor"].str.replace("a","1").replace("l","2").replace("c","3").replace("y","4").replace("f","5").replace("m","6").replace("n","7").replace("p","8").replace("s","9")
<ipython-input-5-67a172f8eb6b>:3: SettingWithCopyWarning: 
A value is trying to be 

Convert the number to integer

In [6]:
df_subset['class'] = df_subset['class'].astype(int)
df_subset['odor'] = df_subset['odor'].astype(int)
df_subset['cap-shape'] = df_subset['cap-shape'].astype(int)

<ipython-input-6-76c44af850a3>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_subset['class'] = df_subset['class'].astype(int)
<ipython-input-6-76c44af850a3>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_subset['odor'] = df_subset['odor'].astype(int)
<ipython-input-6-76c44af850a3>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pan

Show the distribution of data for each of the columns I selected

In [12]:
df_subset.head()

class  odor  cap-shape
0      1     8          3
1      2     1          3
2      2     2          1
3      1     8          3
4      2     7          3

Import scikit-learn neighbors and metrics library

In [80]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn import metrics

Create the numpy X and y parameters

In [64]:
X_odor=df_subset[["odor"]].to_numpy()
y=df_subset["class"]

Create KNeighborsClassifier object

In [65]:
knn = KNeighborsClassifier(n_neighbors=1)

Fill the data and answer into the classifier

In [66]:
knn.fit(X_odor,y)

KNeighborsClassifier(n_neighbors=1)

Make the prediction based on the odor values

In [73]:
y_odor=knn.predict(X_odor)

Evaluate the rate of accuracy by odor

In [75]:
metrics.accuracy_score(y,y_odor)

0.9852289512555391

Create the numpy X by cap-shape

In [76]:
X_capshape=df_subset[["cap-shape"]].to_numpy()

Fill the data and answer into the classifier for cap-shape

In [77]:
knn.fit(X_capshape,y)

KNeighborsClassifier(n_neighbors=1)

Make the prediction based on the cap-shape values

In [78]:
y_capshape=knn.predict(X_capshape)

Evaluate the rate of accuracy by cap-shape

In [79]:
metrics.accuracy_score(y,y_capshape)

0.5642540620384048

###### Conclusion:
The prediction by odor has an accuracy rate of 98.5% much more than the rate by cap-shape. So, using odor is more accurately predicts whether or not a mushroom is poisonous.